### DISCUSSION TITLE
Is it possible to use 'defer' when rollback a transaction?

### DISCUSSION THREAD

**Discussion by yuki2006** [ent/ent#2333] [Q&A]
(The English has been translated. sorry.)

Hi, I recently thought about it.

For example.
If we write as below , when a runtime error occurs rollback is not called.

```
	values := []int{100}

	tx, err := client.Tx(ctx)
	if err != nil {
		return
	}
	_, err = tx.Client().User.Create().
		SetAge(values[1]).   // runtime error
		Save(ctx)
	if err != nil {
                 tx.Rollback()
		return err
	}
	tx.Commit()
```

Here the best practice method is good in that a "rollback" is called.

https://entgo.io/docs/transactions#best-practices

But, for example, in such a case, the variable has to be declared out of scope, so there is no benefit of type inference.

(In a real program, it would be more complicated.)
```

func getUsers() []*ent.User {
	ctx := context.Background()

	values := []int{100}

	var users []*ent.User

	WithTx(ctx, client, func(tx *ent.Tx) error {

		var err error

		_, err = tx.Client().User.Create().
			SetAge(values[0]).
			Save(ctx)
		if err != nil {
			return err
		}

		users, err = tx.Client().User.Query().
			Where(user.IDEQ(values[1])). // runtime error
			All(ctx) 
		if err != nil {
			return err
		}
		return nil
	})
	return users
}
```

So if we can use "defer", we can write it like this and benefit from type inference.

But right now, Even if there are no errors, the debug output says "rollbacked", and I don't know if this is a good way (in the future).

Of course, depending on the situation, the life of the transaction needs to be considered, but is it possible to choose to "defer"?

```
func getUsers() ([]*ent.User, error) {
	ctx := context.Background()

	values := []int{100}

	tx, err := client.Tx(ctx)
	defer tx.Rollback()
	if err != nil {
		return nil, err
	}

	_, err = tx.Client().User.Create().
		SetAge(values[0]).
		Save(ctx)
	if err != nil {
		return nil, err
	}

	users, err := tx.Client().User.Query().
		Where(user.IDEQ(values[1])). // runtime error
		All(ctx)

	if err != nil {
		return nil, err
	}
	tx.Commit()

	return users, nil
}

```

**Comment by crossworth** [comment]
Hello, I think this is already possible.
If I'm not mistaken if you commit the transaction and call rollback, it will not rollback it (it may return an error saying that the transaction is already committed).

The `WithTx` is a very useful helper since your code can be very clean, but you could in theory have the same behaviour by using [Transactional Mutations](https://entgo.io/docs/tutorial-todo-gql-tx-mutation) when using `entgql` or could create a custom middleware that creates a transaction and injects the Ent client (with the transaction binded to it) on the context (using `tx.Client()` `ent.NewContext()` `ent.FromContext()`) and do the rollback/commit.

**Reply by yuki2006** [reply]
thx @crossworth .

My guess is that it's working fine.
However. The fact that there is no such description in the manual, and the fact that the debug output shows "rollback", makes me think that it is better not to use it.

I would like to see it mentioned in the manual, and if it is, I can use it with confidence.
